src/wh_client_she.c

```diff
@@ -0,0 +1,879 @@
+    ret = wh_Client_RecvResponse(c, &group, &action, &dataSz, (uint8_t*)packet);
+    if (ret == 0) {
+        if (packet->rc != WOLFHSM_SHE_ERC_NO_ERROR)
+            ret = packet->rc;
+        else if (sz < packet->sheEncCbcRes.sz)
+            ret = WH_ERROR_BADARGS;
+        else
+            memcpy(out, packOut, packet->sheEncCbcRes.sz);
+    }
+    return ret;
+}
+
+int wh_Client_SheEncCbc(whClientContext* c, uint8_t keyId, uint8_t* iv,
+    uint32_t ivSz, uint8_t* in, uint8_t* out, uint32_t sz)
+{
+    int ret;
+    ret = wh_Client_SheEncCbcRequest(c, keyId, iv, ivSz, in, sz);
+    if (ret == 0) {
+        do {
+            ret = wh_Client_SheEncCbcResponse(c, out, sz);
+        } while (ret == WH_ERROR_NOTREADY);
+    }
+    return ret;
+}
+
+int wh_Client_SheDecEcbRequest(whClientContext* c, uint8_t keyId, uint8_t* in,
+    uint32_t sz)
+{
+    int ret;
+    uint8_t* packIn;
+    whPacket* packet;
+    if (c == NULL || in == NULL || sz < WOLFHSM_SHE_KEY_SZ)
+        return WH_ERROR_BADARGS;
+    packet = (whPacket*)wh_CommClient_GetDataPtr(c->comm);
+    /* in is after fixed sized fields */
+    packIn = (uint8_t*)(&packet->sheDecEcbReq + 1);
+    packet->sheDecEcbReq.keyId = keyId;
+    packet->sheDecEcbReq.sz = sz;
+    /* set in */
+    memcpy(packIn, in, sz);
+    /* send enc ecb */
+    ret = wh_Client_SendRequest(c, WH_MESSAGE_GROUP_SHE, WH_SHE_DEC_ECB,
+        WOLFHSM_PACKET_STUB_SIZE + sizeof(packet->sheDecEcbReq) + sz,
+        (uint8_t*)packet);
+    return ret;
+}
+
+int wh_Client_SheDecEcbResponse(whClientContext* c, uint8_t* out, uint32_t sz)
+{
+    int ret;
+    uint16_t group;
+    uint16_t action;
+    uint16_t dataSz;
+    uint8_t* packOut;
+    whPacket* packet;
+    if (c == NULL || out == NULL || sz < WOLFHSM_SHE_KEY_SZ)
+        return WH_ERROR_BADARGS;
+    packet = (whPacket*)wh_CommClient_GetDataPtr(c->comm);
+    /* out is after fixed sized fields */
+    packOut = (uint8_t*)(&packet->sheDecEcbRes + 1);
+    ret = wh_Client_RecvResponse(c, &group, &action, &dataSz, (uint8_t*)packet);
+    if (ret == 0) {
+        if (packet->rc != WOLFHSM_SHE_ERC_NO_ERROR)
+            ret = packet->rc;
+        else if (sz < packet->sheDecEcbRes.sz)
+            ret = WH_ERROR_BADARGS;
+        else
+            memcpy(out, packOut, packet->sheDecEcbRes.sz);
+    }
+    return ret;
+}
+
+int wh_Client_SheDecEcb(whClientContext* c, uint8_t keyId, uint8_t* in,
+    uint8_t* out, uint32_t sz)
+{
+    int ret;
+    ret = wh_Client_SheDecEcbRequest(c, keyId, in, sz);
+    if (ret == 0) {
+        do {
+            ret = wh_Client_SheDecEcbResponse(c, out, sz);
+        } while (ret == WH_ERROR_NOTREADY);
+    }
+    return ret;
+}
+
+int wh_Client_SheDecCbcRequest(whClientContext* c, uint8_t keyId, uint8_t* iv,
+    uint32_t ivSz, uint8_t* in, uint32_t sz)
+{
+    int ret;
+    uint8_t* packIn;
+    whPacket* packet;
+    if (c == NULL || in == NULL || sz < WOLFHSM_SHE_KEY_SZ || iv == NULL ||
```

Comment:
Not sure I understand why the input size is limited to the key size which is the same as the block size.  Does the user have to repeatedly call this function with each block and make sure the iv is pointed to the previous cipher text?

Reply:
< WOLFHSM_SHE_KEY_SZ, must be at least one block size, multiple blocks are fine

Reply:
Yeah, dumb me!